### PR TITLE
[bazel] Add rules and aspects for dumping toolchain side-products

### PR DESCRIPTION
**Pull request by mcy**
This PR adds some new rules that output assembly files for every TU linked into an `opentitan_binary`, as well as a linker map of the final link. It also adds a rule that will output textual LLVM IR once we switch to clang (which is mostly for my and, ostensibly, Luis's benefit to debug mysterious failures).

This unfortunately has a cost: due to a Bazel limitation, it won't let us add new files to cc_binary or cc_library, so we're forced to run each of those rules twice. If this turns out to be a serious regression, it should be easy to add a Bazel flag for skipping those targets, so that we can disable them in CI.

### CODE REVIEW

**Comment by mcy** [comment]
Fixes https://github.com/lowRISC/opentitan/issues/11381

**Comment by drewmacrae** [comment]
I don't mind if we slow down compilation by a factor of 2 personally. I currently spend most of my time waiting for verilator and rust to build and for long running tests to execute in verilator so it wont have an impact on me or CI AFAIK.

File: rules/opentitan.bzl [hunk 1]
```diff
@@ -120,17 +115,14 @@ sign_bin = rule(
...
-        ),
     },
 )
 
 def _elf_to_disassembly_impl(ctx):
     cc_toolchain = find_cc_toolchain(ctx)
     outputs = []
     for src in ctx.files.srcs:
-        disassembly = ctx.actions.declare_file("{}.dis".format(src.basename))
+        disassembly = ctx.actions.declare_file("{}.elf.s".format(src.basename))
```
Comment (drewmacrae) [inline rules/opentitan.bzl:126]: super nit: (if this is really a thing then don't worry about it unless you're cherrypicking these commits) this change might be in the wrong commit.

File: rules/cc_side_outputs.bzl [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+    provides = [CcSideProductInfo],
+    toolchains = ["@rules_cc//cc:toolchain_type"],
+    incompatible_use_toolchain_transition = True,
+    fragments = ["cpp"],
+    host_fragments = ["cpp"],
+)
+
+def _cc_llvm_aspect_impl(target, ctx):
+    cc_toolchain = find_cc_toolchain(ctx)
+    if cc_toolchain.compiler.find("clang") == -1:
```
[the remark this answers is not in the document]
Comment (mcy) [reply rules/cc_side_outputs.bzl:147]: I think we can leave this for now and revisit if it becomes an issue. Experimentally, gcc does not seem to choke on -emit-llvm and just ignores it...?

Realistically we won't support non-clang compilers eventually anyways.

**Comment by mcy** [comment]
Drew; me neither. I spoke to a former member of the Bazel team extensively internally for how to best achieve this, and this is basically the supported way to do it.

File: rules/util.bzl [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    """
+    Executes an action that consumes `tabby` and renders spaces
+    in place of tabs as if it were displaed in a smart-tabs context
+    with eight-space tabs.
+
+    This is useful for cleaning up tabs left in GCC-style outputs.
+    """
+
+    ctx.actions.run_shell(
+        outputs = [tabby],
```
Comment (cfrantz) [inline rules/util.bzl:17]: `expanded` ?

File: rules/cc_side_outputs.bzl [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+    outputs = []
+    for dep in ctx.rule.attr.deps:
+        if CcSideProductInfo in dep:
+            outputs += dep[CcSideProductInfo].files
+
+    for source_file in translation_units:
+        output_file = ctx.actions.declare_file(
+            # Some source files in the repo are currently pulled in by multiple
+            # rules, and this is allowed in general, although not a good idea.
```
Comment (hlopko) [inline rules/cc_side_outputs.bzl:78]: This is the only crime I witnessed in this PR :)

File: rules/cc_side_outputs.bzl [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+    feature_configuration = cc_common.configure_features(
+        ctx = ctx,
+        cc_toolchain = cc_toolchain,
+        requested_features = ctx.features,
+        unsupported_features = ctx.disabled_features,
+    )
+
+    c_compiler_path = cc_common.get_tool_for_action(
+        feature_configuration = feature_configuration,
+        action_name = ACTION_NAMES.c_compile,
```
Comment (hlopko) [inline rules/cc_side_outputs.bzl:67]: There is in theory a different compiler used for C and for C++, so to be perfectly safe you could `get_tool_for_action` also for `cpp_compile` action.

Comment (mcy) [reply rules/cc_side_outputs.bzl:67]: I don't think that's a thing we actually need to worry about in this project; we don't support anything other than GCC or Clang (soon to be Clang only, in some cases!).

File: rules/cc_side_outputs.bzl [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+    if ctx.rule.kind not in ["cc_library", "cc_binary", "cc_test"]:
+        return CcSideProductInfo(files = [])
+
+    # This filters out both headers and assembly source inputs, neither of which
+    # make sense to generate an .s output for.
+    translation_units = [
+        src.files.to_list()[0]
+        for src in ctx.rule.attr.srcs
+        if src.files.to_list()[0].extension in [
+            # We only use .cc, but to avoid nasty surprises we support all five
```
Comment (drewmacrae) [inline rules/cc_side_outputs.bzl:31]: Nit: there are six elements below

**Review by drewmacrae — APPROVED** [review]

File: rules/cc_side_outputs.bzl [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+            """,
+        )
+
+    return [CcSideProductInfo(files = depset(
+        direct = outputs,
+        transitive = transitive,
+    ))]
+
+def _cc_assembly_aspect_impl(target, ctx):
+    return _cc_compile_different_output(target, ctx, "s", ["-S"])
```
Comment (cfrantz) [inline rules/cc_side_outputs.bzl:159]: TODO:
Emit assembly with original source as comments.
Emit other compiler outputs (e.g. preprocessed source).

Ok to do this in a follow-up PR.

**Review by cfrantz — APPROVED** [review]

**Pull request merged**